WebAPI/Controllers/CategoryController.cs

```diff
@@ -8,54 +8,64 @@
 using WebAPI.Services;
 using AutoMapper;
 using WebAPI.DTOs;
+using WebAPI.Interfaces;
 
 namespace WebAPI.Controllers
 {
     [Route("api/[controller]")]
     [ApiController]
     public class CategoryController : ControllerBase
     {
-        private readonly ICategoryService categoryService;
-        public CategoryController(ICategoryService categoryService)
+        private readonly ICreatorService<Category> _creator;
+        private readonly IGetterService<Category> _getter;
+        private readonly IUpdaterService<Category> _updater;
+        private readonly IDeleterService<Category> _deleter;
+
+        public CategoryController(ICreatorService<Category> creator, IGetterService<Category> getter, IUpdaterService<Category> updater, IDeleterService<Category> deleter)
         {
-            this.categoryService = categoryService;
+            _creator = creator;
+            _getter = getter;
+            _updater = updater;
+            _deleter = deleter;
         }
 
         [HttpGet("{id:int}")]
         public async Task<ActionResult> GetCategory(int id)
         {
-            var result = await categoryService.GetCategory(id);
+            var result = await _getter.GetById<CategoryResponseDto>(id);
             if (result == null) return NotFound();
             return Ok(result);
         }
         [HttpGet]
         public async Task<ActionResult> GetCategories()
         {
-            return Ok(await categoryService.GetCategories());
+            var result = await _getter.GetAll<CategoryResponseDto>();
+            return Ok(result);
         }
 
         [HttpPost]
-        public async Task<ActionResult> Create(CreateCategoryDTO category)
+        public async Task<ActionResult> Create(CategoryRequestDto category)
         {
             if (category == null) return BadRequest();
-            var newCategory = await categoryService.CreateCategory(category);
-            return CreatedAtAction(nameof(GetCategory), new { id = newCategory.id }, newCategory);
+            var newCategory = await _creator.Create<CategoryResponseDto>(category);
+            return Created($"api/category/{newCategory.Id}", newCategory);
         }
 
+
         [HttpPut("{id:int}")]
-        public async Task<ActionResult> Edit(int id, EditCategoryDTO category)
+        public async Task<ActionResult> Edit(int id, CategoryRequestDto category)
         {
-            if (categoryService.GetCategory(id).Result == null) return NotFound();
-            var result = await categoryService.EditCategory(id, category);
-            return Ok(result);
+            if (_getter.GetById<CategoryResponseDto>(id).Result == null) return NotFound();
+            await _updater.Update(category, id);
```

Comment:
Maybe the updater should check if entity exists, instead of checking it in controller? Then you can use if logic based on task result. Also, if needed, you can always extend this updater with your own logic for this particular entity

Reply:
UpdaterService throws a NotFoundException when checking if an entity exists, is this fine, or should I create an UpdaterService extension class? Justyn says not to modify the base one as it might cause problems later on

Reply:
If so, consider using try{} catch{}